### PR TITLE
[APIView] Bug: Automatic revision cleanup leaves stale pending revisions behind

### LINKED ISSUE #15238: [APIView] Bug: Automatic revision cleanup leaves stale pending revisions behind
`CreateAutomaticRevisionAsync` maintains an invariant: only one pending (unreviewed) automatic revision should exist since the last approval. Two bugs in the cleanup logic violated this invariant.

**Bug 1 — Cleanup loop exits early on content match**

The loop that deletes stale pending revisions stopped processing as soon as it found a revision whose content matched the incoming build. Any older pending revisions in the list were never visited and never deleted. This is wrong — the loop's job is cleanup, not content comparison. Content comparison should only happen *after* cleanup is complete.

**Bug 2 — Stale candidate not deleted before creating a new revision**

When the newest pending revision (the "candidate") does not match the incoming content, a new revision needs to be created. However, the candidate was not being deleted before the new revision was created. This left two pending revisions alive at the same time — the stale candidate and the newly created one.

Both bugs have been present since the file was first introduced.

### Fix

- Removed content comparison from inside the cleanup loop entirely. The loop now only performs cleanup; the candidate is always the newest unprotected revision.
- Added deletion of the unprotected candidate before creating a new revision when the content doesn't match.

### PR DESCRIPTION
**Pull request by AlitzelMendez**
closes: https://github.com/Azure/azure-sdk-tools/issues/15238

# Changes explained 

## Before

Every time a CI or release pipeline uploaded a new API revision, the service looped through **all** existing automatic revisions from oldest to newest and soft-deleted every one that didn't match the incoming content — stopping only when it found a content match. This had two problems:

1. **Bulk deletion risk.** If a package had accumulated many pending revisions (e.g. from several CI builds while the PR queue was backed up), a single new upload would wipe all of them in one call. A reviewer who had left comments on any of those revisions would lose that context silently, because the loop didn't protect commented revisions.

2. **Rolling builds always created a new revision.** Nightly/alpha builds increment a date in the version string (`1.4.0-alpha.20201211` → `1.4.0-alpha.20201212`) even when the API surface is unchanged. The old code considered the package version string when deciding whether to reuse, so these builds were never considered "the same" — accumulating a new revision every single day.

## After

The logic is simplified and made predictable:

- **One candidate, always replaced.** The service picks only the single newest unprotected automatic revision (skipping approved, released, or commented ones) and soft-deletes it. Everything else is left untouched. Older stale revisions are handled by a separate one-time cleanup script (#15239) rather than silently in-band.

- **No reuse — ever.** The "same content, skip creation" path has been removed. Every upload always produces a fresh revision. This makes the behaviour easy to reason about and avoids the class of bugs where a stale revision was returned as if it were the newly uploaded one.

- **Version scoping is applied once, up front.** Before selecting the candidate, all automatic revisions are filtered to those belonging to the same *logical version* as the incoming upload (using the normalised `APIVersionId`). This means a `1.0.0` upload never touches `2.0.0-beta` revisions, and vice versa, regardless of which code path runs next.

- **Release pipeline path unchanged.** The `compareAllRevisions` fast-path (used during release to detect whether a newly uploaded stable version is already approved) still works exactly as before.

### CODE REVIEW

File: src/dotnet/APIView/APIViewUnitTests/AutoReviewServiceTests.cs [hunk 1]
```diff
@@ -745,10 +745,10 @@ public async Task CreateAutomaticRevisionAsync_WhenAllPendingRevisionsDeleted_Cr
         }
 
         [Fact]
-        public async Task CreateAutomaticRevisionAsync_WhenLastRevisionDeletedButMatchesWithVersion_CreatesNewRevision()
+        public async Task CreateAutomaticRevisionAsync_WhenRevisionMatchesWithVersionConsideration_ReusesIt()
         {
-            // This test covers Edge Case 2: Comparison parameter inconsistency
-            // Without version: different (deleted), with version: same (would match after deletion)
+            // The reuse check uses considerPackageVersion=true when the incoming code file has a version.
+            // A revision that matches with version should be reused, not deleted and recreated.
```
Comment (tjprescott) [inline src/dotnet/APIView/APIViewUnitTests/AutoReviewServiceTests.cs]: So what was the issue? It was doing this "delete and create" thing but not actually deleting?

File: src/dotnet/APIView/APIViewUnitTests/AutoReviewServiceTests.cs [hunk 2]
```diff
@@ -953,6 +931,434 @@ public async Task CreateAutomaticRevisionAsync_SupersessionScopedToSameVersionId
...
+        ///   B1 — approved                    (anchor — must never be touched)
+        ///
+        /// Expected outcome when B5 (also same content as B4) arrives:
+        ///   - B4 is kept as the candidate (newest unprotected)
+        ///   - B3 is deleted  (older unprotected)
+        ///   - B2 is deleted  (older unprotected)
+        ///   - B1 is NOT deleted (approved anchor — hard stop)
+        ///   - B5 is NOT created (B4 content matches incoming)
+        ///   - Returned revision is B4
+        /// </summary>
```
Comment (tjprescott) [inline src/dotnet/APIView/APIViewUnitTests/AutoReviewServiceTests.cs]: Okay, so this both will not create a new revision for B5, but it will also delete any existing duplicates?  Are we sure we want to go that far? We could deal with existing duplicates with the cleanup script in https://github.com/Azure/azure-sdk-tools/issues/15239. If the dedup logic is working, then the duplicates would never get created in the first place.

File: src/dotnet/APIView/APIViewUnitTests/AutoReviewServiceTests.cs [hunk 3]
```diff
@@ -953,6 +931,434 @@ public async Task CreateAutomaticRevisionAsync_SupersessionScopedToSameVersionId
...
+        ///   B4 — pending, no comments
+        ///   B3 — pending, HAS a comment  ← protected
+        ///   B2 — pending, no comments
+        ///   B1 — approved anchor
+        ///
+        /// Expected outcome when B5 (any content) arrives:
+        ///   - B4 is kept as candidate
+        ///   - B3 is hit: protected by comment → hard stop, B2 is never visited
+        ///   - B2 and B1 are NOT deleted
+        /// </summary>
```
Comment (tjprescott) [inline src/dotnet/APIView/APIViewUnitTests/AutoReviewServiceTests.cs]: I find this confusing. Not sure why they are B instead of R (since they are revisions) but regardless, are these revisions the same, different?

Comment (tjprescott) [reply src/dotnet/APIView/APIViewUnitTests/AutoReviewServiceTests.cs]: Separately, I wonder if our Version-centric proposal will adapt correctly?

File: src/dotnet/APIView/APIViewUnitTests/AutoReviewServiceTests.cs [hunk 4]
```diff
@@ -953,6 +931,434 @@ public async Task CreateAutomaticRevisionAsync_SupersessionScopedToSameVersionId
...
+        [Fact]
+        public async Task CreateAutomaticRevisionAsync_MultiBuildProgression_StopsAtCommentedRevision()
```
Comment (tjprescott) [inline src/dotnet/APIView/APIViewUnitTests/AutoReviewServiceTests.cs]: Why does it stop at a commented revision?

File: src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs [hunk 5]
```diff
@@ -83,31 +83,34 @@ public AutoReviewService(
...
-                    while (automaticRevisionsQueue.Count > 0)
-                    {
-                        latestAutomaticAPIRevision = automaticRevisionsQueue.Dequeue();
 
-                        // Check if we should keep this revision
-                        if (latestAutomaticAPIRevision.IsApproved ||
-                            latestAutomaticAPIRevision.IsReleased ||
-                            await _apiRevisionsManager.AreAPIRevisionsTheSame(latestAutomaticAPIRevision, renderedCodeFile, incomingContentHash: incomingContentHash) ||
-                            comments.Any(c => latestAutomaticAPIRevision.Id == c.APIRevisionId))
+                    foreach (var revision in automaticRevisions)
```
Comment (tjprescott) [inline src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs]: Are we sure this is sorted?

File: src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs [hunk 6]
```diff
@@ -83,31 +83,34 @@ public AutoReviewService(
...
+                    {
+                        // Hard stop: this revision and everything older is protected — stop processing
+                        if (revision.IsApproved || revision.IsReleased || revisionIdsWithComments.Contains(revision.Id))
```
Comment (tjprescott) [inline src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs]: Why is revision with comment a hard stop?

File: src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs [hunk 7]
```diff
@@ -83,31 +83,34 @@ public AutoReviewService(
...
                             break;
                         }
 
-                        // Delete this revision
-                        await _apiRevisionsManager.SoftDeleteAPIRevisionAsync(apiRevision: latestAutomaticAPIRevision, notes: "Deleted by Automatic Review Creation...");
-                        latestAutomaticAPIRevision = null;  // Mark as consumed
+                        if (latestAutomaticAPIRevision == null)
+                        {
+                            // First (newest) unprotected revision: keep it as the candidate to reuse or replace
+                            latestAutomaticAPIRevision = revision;
```
Comment (tjprescott) [inline src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs]: Is there a difference between "reuse" and "replace"?

File: src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs [hunk 8]
```diff
@@ -83,31 +83,34 @@ public AutoReviewService(
...
+                        }
+                        else
+                        {
+                            // Older unprotected revision: always delete
+                            await _apiRevisionsManager.SoftDeleteAPIRevisionAsync(apiRevision: revision, notes: "Deleted by Automatic Review Creation...");
```
Comment (tjprescott) [inline src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs]: I'm not sure we need to go this far. We could do this in a cleanup. The idea is just to see if the same revision essentially already exists, yes? And then to avoid creating a duplicate.

File: src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs [hunk 9]
```diff
@@ -83,31 +83,34 @@ public AutoReviewService(
...
+                        }
                     }
 
-                    // We should compare against only latest revision when calling this API from scheduled CI runs
-                    // But any manual pipeline run at release time should compare against all approved revisions to ensure hotfix release doesn't have API change
-                    // If review surface doesn't match with any approved revisions then we will create new revision if it doesn't match pending latest revision
+                    // For release pipeline runs, compare against all approved revisions first to catch hotfix API changes.
+                    // Otherwise, check if the candidate revision (newest unprotected) matches the incoming content.
+                    // If it matches, reuse it. If not, delete the candidate and create a new revision.
```
Comment (tjprescott) [inline src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs]: Well, if there's a hotfix (with no API change, just a version change) we *do* want a new revision yes? Don't we need a new revision to release it?

File: src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs [hunk 10]
```diff
@@ -129,6 +132,14 @@ await _apiRevisionsManager.AreAPIRevisionsTheSame(latestAutomaticAPIRevision, re
                         apiRevision = latestAutomaticAPIRevision;
                         createNewRevision = false;
                     }
+                    else if (latestAutomaticAPIRevision != null &&
+                             !latestAutomaticAPIRevision.IsApproved &&
+                             !latestAutomaticAPIRevision.IsReleased &&
+                             !revisionIdsWithComments.Contains(latestAutomaticAPIRevision.Id))
+                    {
+                        // Candidate doesn't match incoming content and is unprotected — delete it before creating the new revision
+                        await _apiRevisionsManager.SoftDeleteAPIRevisionAsync(apiRevision: latestAutomaticAPIRevision, notes: "Deleted by Automatic Review Creation...");
```
Comment (tjprescott) [inline src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs]: Okay, so this is replace... the new content matches in every tracked way, but maybe not in an untracked way (i.e. something in skipDiff tags changed) so we delete and create. That make sense to me. 

So in that case, when would we "reuse"

**Review by tjprescott — COMMENTED** [review]
Mostly just questions. I do wonder if this PR goes further than needed.

File: src/dotnet/APIView/APIViewUnitTests/AutoReviewServiceTests.cs (hunk 1 above)
Comment (AlitzelMendez) [reply src/dotnet/APIView/APIViewUnitTests/AutoReviewServiceTests.cs]: exactly!! it was doing the "AreAPIRevisonTheSame" on the cleanup loop and it was used as stop condition, as soon as the content matched it broke the loop! but the one that was the same was never deleted!!!

I mean this is the main issue :) we can stick to use fixing this part :) and don't visit the older revisions as before!

File: src/dotnet/APIView/APIViewUnitTests/AutoReviewServiceTests.cs (hunk 2 above)
Comment (AlitzelMendez) [reply src/dotnet/APIView/APIViewUnitTests/AutoReviewServiceTests.cs]: yes! this also makes sense!!!

File: src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs (hunk 5 above)
Comment (AlitzelMendez) [reply src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs]: yes, they are sorted on line 69                
` apiRevisions = apiRevisions.OrderByDescending(r => r.CreatedOn);`
they are filtered again on line 80, but from the sorted one :)

File: src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs (hunk 6 above)
Comment (AlitzelMendez) [reply src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs]: deleting a revision with reviewer comments would silently destroy those comments, like not really destroy them but the comments are attached by APIRevisionId so if the revision is gone, the comments become orphaned/invisible, probably we will be able to fix this with the APIVersionId 

but for now it will happen like when you encounter orphan comments

File: src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs (hunk 7 above)
Comment (AlitzelMendez) [reply src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs]: reuse = the candidate's content matches incoming → return it as-is, no deletion, no new revision created.
replace = the candidate's content doesn't match → delete the candidate, create a new revision with the incoming content.

as we just one pending

File: src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs (hunk 8 above)
Comment (AlitzelMendez) [reply src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs]: yes we could, the original intention was to do this, at least from the comments I think, but also looks scary so let me change the approach and ping you after that :D

File: src/dotnet/APIView/APIViewUnitTests/AutoReviewServiceTests.cs (hunk 3 above)
Comment (AlitzelMendez) [reply src/dotnet/APIView/APIViewUnitTests/AutoReviewServiceTests.cs]: yes, it should, we are already using the version centric approach for the automatic revisions filtering :) 

right now, that most of the cases are landing in `string.IsNullOrEmpty(r.APIVersionId)` but that will change

```chsarp
var automaticRevisions = apiRevisions.Where(r => r.APIRevisionType == APIRevisionType.Automatic && 
(incomingVersionModel == null|| r.APIVersionId == incomingVersionModel.Id || string.IsNullOrEmpty(r.APIVersionId)))
```

File: src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs (hunk 9 above)
Comment (AlitzelMendez) [reply src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs]: yes!! considerPackageVersion consider this 

```  if (await _apiRevisionsManager.AreAPIRevisionsTheSame(approvedAPIRevision, renderedCodeFile, considerPackageVersion, incomingContentHash))```

so as the version is different a new one is created :)

File: src/dotnet/APIView/APIViewUnitTests/AutoReviewServiceTests.cs (hunk 2 above)
Comment (AlitzelMendez) [reply src/dotnet/APIView/APIViewUnitTests/AutoReviewServiceTests.cs]: ok simplified and now just deleted the newest unprotected

File: src/dotnet/APIView/APIViewUnitTests/AutoReviewServiceTests.cs (hunk 4 above)
Comment (AlitzelMendez) [reply src/dotnet/APIView/APIViewUnitTests/AutoReviewServiceTests.cs]: well now it doesn't stop but it skips it and is to avoid the ghost comments

File: src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs (hunk 10 above)
Comment (AlitzelMendez) [reply src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs]: same content, same version, we resuse previous one instead of creating a new revision, like of the every day rolling build with no changes between them, we should stop creating a new revision every single day

Comment (tjprescott) [reply src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs]: Yeah, but even then, they can have the same version and same hash, but if the "skipDiff" content is different, we would want to replace instead of reuse, no?

Comment (AlitzelMendez) [reply src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs]: yes! makes sense!! done :)

File: src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs (hunk 8 above)
Comment (AlitzelMendez) [reply src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs]: done!

File: src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs (hunk 7 above)
Comment (AlitzelMendez) [reply src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs]: not anymore! no more reuse,

**Review by tjprescott — APPROVED** [review]
Would be good to get @praveenkuttappan's review too!

File: src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs [hunk 11]
```diff
@@ -122,12 +97,16 @@ await _apiRevisionsManager.AreAPIRevisionsTheSame(latestAutomaticAPIRevision, re
...
                     }
 
-                    // Only reuse latestAutomaticAPIRevision if one was kept
-                    if (latestAutomaticAPIRevision != null &&
-                        await _apiRevisionsManager.AreAPIRevisionsTheSame(latestAutomaticAPIRevision, renderedCodeFile, considerPackageVersion, incomingContentHash))
+                    var comments = await _commentsManager.GetCommentsAsync(review.Id);
+                    var revisionIdsWithComments = comments.Select(c => c.APIRevisionId).ToHashSet();
+
+                    // Find the newest pending automatic revision to replace.
+                    var latestAutomaticAPIRevision = automaticRevisions.FirstOrDefault(
```
Comment (praveenkuttappan) [inline src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs:103]: I can think of few issues with this approach when a team run CI pipeline from a different branch for a hotfix.

Comment (AlitzelMendez) [reply src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs:103]: great point! fixed

File: src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs [hunk 12]
```diff
@@ -102,7 +101,8 @@ public AutoReviewService(
 
                     // Find the newest pending automatic revision to replace.
                     var latestAutomaticAPIRevision = automaticRevisions.FirstOrDefault(
-                        r => !r.IsApproved && !r.IsReleased && !revisionIdsWithComments.Contains(r.Id));
+                        r => !r.IsApproved && !r.IsReleased && !revisionIdsWithComments.Contains(r.Id)
+                        && r.PackageVersion == codeFile.PackageVersion);
```
Comment (tjprescott) [inline src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs:105]: `codeFile.PackageVersion` implies to me that it's the version from the code file, which would be the wonky alpha version. Is that the case or is it actually comparing the resolved version labels?

Comment (AlitzelMendez) [reply src/dotnet/APIView/APIViewWeb/Managers/AutoReviewService.cs:105]: we replace that value right away to avoid having to track it and consider it everywhere 
https://github.com/Azure/azure-sdk-tools/blob/cca847c9744ad0d03f9d6c3fa15b91ddea6d48eb/src/dotnet/APIView/APIViewWeb/LeanControllers/AutoReviewController.cs#L77

(that is what was happening at the beginning before I fixed it :O)

**Review by tjprescott — APPROVED** [review]
